src/Microsoft.Extensions.Diagnostics.HealthChecks/HealthCheckService.cs

```diff
@@ -6,127 +6,103 @@
-
-        /// <summary>
-        /// Constructs a <see cref="HealthCheckService"/> from the provided collection of <see cref="IHealthCheck"/> instances, and the provided logger.
-        /// </summary>
-        /// <param name="healthChecks">The <see cref="IHealthCheck"/> instances that have been registered in the application.</param>
-        /// <param name="logger">A <see cref="ILogger{T}"/> that can be used to log events that occur during health check operations.</param>
-        public HealthCheckService(IEnumerable<IHealthCheck> healthChecks, ILogger<HealthCheckService> logger)
+        public HealthCheckService(IServiceScopeFactory scopeFactory, ILogger<HealthCheckService> logger)
         {
-            healthChecks = healthChecks ?? throw new ArgumentNullException(nameof(healthChecks));
+            _scopeFactory = scopeFactory ?? throw new ArgumentNullException(nameof(scopeFactory));
             _logger = logger ?? throw new ArgumentNullException(nameof(logger));
 
-            // Scan the list for duplicate names to provide a better error if there are duplicates.
-            var names = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
-            var duplicates = new List<string>();
-            foreach (var check in healthChecks)
+            // We're specifically going out of our way to do this at startup time. We want to make sure you
+            // get any kind of health-check related error as early as possible. Waiting until someone
+            // actually tries to **run** health checks would be real baaaaad.
+            using (var scope = _scopeFactory.CreateScope())
             {
-                if (!names.Add(check.Name))
-                {
-                    duplicates.Add(check.Name);
-                }
+                var healthChecks = scope.ServiceProvider.GetRequiredService<IEnumerable<IHealthCheck>>();
+                EnsureNoDuplicates(healthChecks);
             }
+        }
 
-            if (duplicates.Count > 0)
+        public Task<CompositeHealthCheckResult> CheckHealthAsync(CancellationToken cancellationToken = default) =>
+            CheckHealthAsync(predicate: null, cancellationToken);
+
+        public async Task<CompositeHealthCheckResult> CheckHealthAsync(
+            Func<IHealthCheck, bool> predicate,
+            CancellationToken cancellationToken = default)
+        {
+            using (var scope = _scopeFactory.CreateScope())
             {
-                throw new ArgumentException($"Duplicate health checks were registered with the name(s): {string.Join(", ", duplicates)}", nameof(healthChecks));
-            }
+                var healthChecks = scope.ServiceProvider.GetRequiredService<IEnumerable<IHealthCheck>>();
 
-            Checks = healthChecks.ToDictionary(c => c.Name, StringComparer.OrdinalIgnoreCase);
+                if (predicate != null)
+                {
+                    healthChecks = healthChecks.Where(predicate);
```

Comment:
You could avoid the Linq call by doing this inside the `foreach`

Reply:
sure